File: leetcode/200-number-of-islands/number-of-islands.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {

        int compo = 0;
        int row = grid.size() , col = grid[0].size();
        vector<vector<bool>> vis(row,vector<bool>(col,false));
        for(int i = 0 ; i < row;i++){
            for(int j = 0; j < col ; j++){
                 if(grid[i][j] == '1' && vis[i][j] == false){
                    compo++;
                    dfs(grid,vis,i,j,row,col);
                 }
            }
        }

         return compo;       
    }

    void dfs(vector<vector<char>>& grid , vector<vector<bool>>& vis , int i , int j,int row, int col){
          if(i < 0 || i == row || j < 0 || j == col || grid[i][j] == '0' || vis[i][j] == true)
              return;

          vis[i][j] = true;

          dfs(grid,vis,i-1,j,row,col);
          dfs(grid,vis,i+1,j,row,col);
          dfs(grid,vis,i,j-1,row,col);
          dfs(grid,vis,i,j+1,row,col);

          return;
    }
};

File: leetcode/200-number-of-islands/number-of-islands.java
class Solution {
    public int numIslands(char[][] grid) {

        boolean[][] vis = new boolean[grid.length][grid[0].length];
        int cnt = 0;
        for(int i = 0 ; i < grid.length ; i++){
            for(int j = 0 ; j < grid[0].length; j++){
                 if(grid[i][j] == '1' && vis[i][j] == false){
                    cnt++;
                    dfs(grid,vis,i,j);
                 }
            }
        }
        return cnt;
        
    }

    public void dfs(char[][] grid,boolean[][] vis,int i,int j){
        if(i < 0 || i >= grid.length || j < 0 || j >= grid[0].length || vis[i][j] == true || grid[i][j] == '0')
            return;

        vis[i][j] = true;
        dfs(grid,vis,i-1,j);
        dfs(grid,vis,i+1,j);
        dfs(grid,vis,i,j-1);
        dfs(grid,vis,i,j+1);
    }
}